File: 200.岛屿数量.java
/*
 * @lc app=leetcode.cn id=200 lang=java
 *
 * [200] 岛屿数量
 */

// @lc code=start
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;
        int col = grid.length, row = grid[0].length;
        for (int i = 0; i < col; i++) {
            for (int j = 0; j < row; j++) {
                if (grid[i][j] == '1') {
                    dfs(grid, i, j);
                    count++;
                }
            }
        }
        return count;
    }

    public void dfs(char[][] grid, int i, int j) {
        int col = grid.length, row = grid[0].length;
        if (i < 0 || j < 0 ||
            i >= col || j >= row ||
            grid[i][j] == '0') {
                return;
            }
        grid[i][j] = '0';
        dfs(grid, i + 1, j);
        dfs(grid, i, j + 1);
        dfs(grid, i - 1, j);
        dfs(grid, i, j - 1);
    }
}
// @lc code=end

File: 200.岛屿数量.cpp
#include<vector>
#include<queue>
using namespace std;

/*
 * @lc app=leetcode.cn id=200 lang=cpp
 *
 * [200] 岛屿数量
 */

// @lc code=start
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int res = 0;
        int m = grid.size(),
            n = grid[0].size();
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    bfs(grid, i, j);
                    res ++;
                }
            }
        }

        return res;
    }

    void bfs(vector<vector<char>>& nums, int i, int j) {
        queue<pair<int, int>> q;
        q.push(make_pair(i, j));

        while (!q.empty()) {
            auto node = q.front();
            q.pop();
            int row = node.first;
            int col = node.second;

            if (row < 0 || col < 0 || row >= nums.size() || col >= nums[0].size() || nums[row][col] == '0') 
                continue;
            
            nums[row][col] = '0';
            q.push(make_pair(row + 1, col));
            q.push(make_pair(row, col + 1));
            q.push(make_pair(row, col - 1));
            q.push(make_pair(row - 1, col));

        }
    }
};
// @lc code=end